Refactor scheduling logic

ChocoChipset opened the pull request:
- Move all timer logic to `SPTPersistentCacheTimerProxy`.
  
  Encapsulate all logic regarding timing to the `SPTPersistentCacheTimerProxy` class.
  Also unit test it. ;)
- Rename `SPTPersistentCacheTimerProxy` to `SPTPersistentCacheGarbageCollectorScheduler`
  
  After moving the timer logic to `SPTPersistentCacheTimerProxy`, this class is no longer a timer proxy, but an object in charge of scheduling the garbage collection. Hence, rename it to the more descriptive name `SPTPersistentCacheGarbageCollectorScheduler`.

Commit:
Move all timer logic to SPTPersistentCacheTimerProxy

Encapsulate all logic regarding timing to the SPTPersistentCacheTimerProxy class.
Also unit test it. ;)

Commit:
Rename SPTPersistentCacheTimerProxy to SPTPersistentCacheGarbageCollectorScheduler

After moving the timer logic to SPTPersistentCacheTimerProxy, this class is no longer a timer proxy, but an object in charge of scheduling the garbage collection. Hence, rename it to the more descriptive name SPTPersistentCacheGarbageCollectorScheduler.

ChocoChipset commented:
Review: @spotify/objc-dev

rastersize commented on Sources/SPTPersistentCache.m:
```diff
@@ -92,6 +91,11 @@ - (instancetype)initWithOptions:(SPTPersistentCacheOptions *)options
     }
+    
+    _garbageCollectorScheduler = [[SPTPersistentCacheGarbageCollectorScheduler alloc]
+                 initWithDataCache:self
```
The indentation here looks a bit messy :)

8W9aG replied:
:+1: Let's keep it in line with the rest of the project

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,133 @@
+
+#import "SPTPersistentCache+Private.h"
+
+const NSTimeInterval SPTPersistentCacheTimerProxyTimerToleranceInterval = 300;
```
Looks like we could make this `static` as well?

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,133 @@
+
+- (void)dealloc
+{
+    NSTimer *timer = self.timer;
```
Use ivar access.

rastersize replied:
Actually, with this change. Where the proxy has been removed. This object will _never_ be deallocated unless `-unscheduleGarbageCollection` has already been ran. And as such there’s no timer instance.

That is because the timer will hold a strong pointer to our object, which holds a strong pointer to the timer. Very retain-cycle.

So I think you’ll need to either make sure that pre-condition is guaranteed to happen (i.e. someone calls unschedule externally). Or add a proxy object for the timer that hold a weak reference, like the old proxy.

8W9aG replied:
Also the timer will retain this object, making the timer never valid when it gets to dealloc, its better off to have an "invalidate" method on this object which calls invalidate on the timer

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,133 @@
+    _options = options;
+    _dataCache = dataCache;
+    _queue = queue;
+    _debugOutput = [options.debugOutput copy];
```
Why do we need to make a copy of this block when we also have a reference to it via `self.options.debugOutput`?

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,133 @@
+{
+    NSTimer *timer = self.timer;
+    
+    dispatch_async(dispatch_get_main_queue(), ^{
```
Could we run this sync if we’re already on the main queue/thread?

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,133 @@
+
+- (void)scheduleGarbageCollection
+{
+    assert([NSThread isMainThread]);
```
Change to

``` objc
NSAssert([NSThread isMainThread], @"Must be called on the main thread");
```

Or alternatively:

``` objc
if (![NSThread isMainThread]) {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self scheduleGarbageCollection];
    }
}
```

I think I’d prefer if we took the second approach.

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.h:
```diff
@@ -35,14 +36,28 @@
-- (void)enqueueGC:(NSTimer *)timer;
+/**
+ *  Schedules the garbage collection operation.
+ */
```
Iff you keep the assertion, also document that this _have to_ be called on the main thread.

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.h:
```diff
@@ -35,14 +36,28 @@
+
+/**
+ *  Unschedules the garbage collection operation.
+ */
```
Iff you keep the assertion, also document that this _have to_ be called on the main thread.

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,133 @@
+
+- (void)unscheduleGarbageCollection
+{
+    assert([NSThread isMainThread]);
```
Same [note as for the `scheduleGarbageCollection` method](https://github.com/spotify/SPTPersistentCache/pull/19/files#r53555667).

8W9aG commented on Sources/SPTPersistentCache.m:
```diff
@@ -507,14 +486,6 @@ - (NSUInteger)lockedItemsSizeInBytes
     return size;
 }
 
-- (void)dealloc
```
I think we will still need to invalidate the garbage collector scheduler here

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,133 @@
+
+#pragma mark - Debug Logs
+
+- (void)debugOutput:(NSString *)format, ... NS_FORMAT_FUNCTION(1,2)
```
This method [already exists in the file manager](https://github.com/spotify/SPTPersistentCache/blob/9f8b2cf02b60dc87246b6f8b08c59f1fbccceb61/Sources/SPTPersistentCacheFileManager.m#L210-L219). Could we move that to a shared (private) location instead in the spirit of DRYness?

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,133 @@
+{
+    va_list list;
+    va_start(list, format);
+    NSString * str = [[NSString alloc ] initWithFormat:format arguments:list];
```
:art: I know you just moved this but this line should read:

``` objc
NSString *debugDescription = [[NSString alloc] initWithFormat:format arguments:list];
```

I.e. a few extraneous spaces and also let’s give the variable a better name.

Commit:
Fix comments from review for SPTPersistentCacheGarbageCollectorScheduler

Commit:
Merge branch 'master' of github.com:spotify/SPTPersistentCache into timer-refactor

Commit:
Remove some new lines in SPTPersistentCacheTimerProxyTests

Commit:
Abstract debug callback to its own function

Commit:
Rename SPTPersistentCacheTimerProxyTests class

Rename SPTPersistentCacheTimerProxyTests to SPTPersistentCacheGarbageCollectorSchedulerTests

Commit:
Unit test integration between scheduler and cache

Just ensure the scheduler is being activated/deactivated.

ChocoChipset commented:
Fixed all comments.

Didn't think about the possibility of deallocating the scheduler without it being unscheduled first (and causing a retain cycle). 

To solve this, `SPTPersistentCache` unschedules this scheduler on `-dealloc`.

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,141 @@
+
+BOOL SPTPersistentCacheGarbageCollectorSchedulerIsInMainQueue(void);
+
+static const NSTimeInterval SPTPersistentCacheTimerProxyTimerToleranceInterval = 300;
```
Rename variable to `SPTPersistentCacheGarbageCollectorSchedulerTimerTolerance`

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,141 @@
+#import "SPTPersistentCacheTypeUtilities.h"
+#import "SPTPersistentCache+Private.h"
+
+BOOL SPTPersistentCacheGarbageCollectorSchedulerIsInMainQueue(void);
```
Should be made `static`.

Commit:
Made SPTPersistentCacheGarbageCollectorSchedulerIsInMainQueue function static

Also update name of constant used in SPTPersistentCacheGarbageCollectorScheduler.

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,141 @@
+
+- (void)dealloc
+{
+    NSTimer *timer = _timer;
```
Sorry, the retain-cycle problem is still here unfortunately in the generic case. As if the timer hasn’t been invalidated you’ll never reach this method. And if it has been invalidated already your implementation doesn’t do anything.

So do **one** of the following two solutions. Of which **I’d prefer option number two** as it’s the most resilient with the least likelihood of introducing retain-cycles due to human error.

## Solution one, let external users deal with it

Document that `-unscheduleGarbageCollection` must be sent to the scheduler or it will lead to horrors. Then remove the entire `-dealloc` implementation as it will never be called when a timer is alive, and as such it can’t invalidate said timer.

## Solution two, add a weak proxy to deal with it automatically

Add a `NSProxy` subclass called `SPTPersistentCacheWeakTargetProxy` and use it as the timer’s target:

``` objc
- (void)scheduleGarbageCollection
{
    // ...

    self.timer = [NSTimer timerWithTimeInterval:self.options.gcIntervalSec
                                         target:[SPTPersistentCacheWeakTargetProxy proxyWithTarget:self]
                                       selector:@selector(enqueueGarbageCollection:)
                                       userInfo:nil
                                        repeats:YES];

    // ...
}
```

The implementation of `SPTPersistentCacheWeakTargetProxy` would look like:

``` objc
// sources/SPTPersistentCacheWeakTargetProxy.h

NS_ASSUME_NONNULL_BEGIN

/**
 * A proxy class which insulates it’s target from being strongly retained. It will forward any messages sent
 * to itself to its target.
 *
 * Can be used together with `NSTimer` to make sure a retain-cycle isn’t created.
 */
@interface SPTPersistentCacheWeakTargetProxy : NSProxy

/**
 * Create a new weak target proxy for the given _target_.
 *
 * @param target The target to which all received messages should be forwarded.
 */
+ (instancetype)proxyWithTarget:(id)target;

@end

NS_ASSUME_NONNULL_END


// sources/SPTPersistentCacheWeakTargetProxy.m
@interface SPTPersistentCacheWeakTargetProxy ()

@property (nonatomic, weak, readonly) id target;

@end

@implementation SPTPersistentCacheWeakTargetProxy

#pragma mark Object Life Cycle

+ (instancetype)proxyWithTarget:(id)target
{
    return [[self alloc] initWithTarget:target];
}

- (instancetype)initWithTarget:(id)target
{
    _target = target;
    return self;
}

#pragma mark Forwarding Messages

- (id)forwardingTargetForSelector:(SEL)aSelector
{
    return self.target;
}

- (NSMethodSignature *)methodSignatureForSelector:(SEL)selector
{
    id target = self.target;

    NSMethodSignature *methodSignature;

    if (target == nil) {
        methodSignature = [target methodSignatureForSelector:selector];
    } else {
        // We still want to report to the runtime that we found a selector. It’ll just be sent to `nil` later
        // or not at all, so not a problem that it’s fake.
        NSString * const types = [NSString stringWithFormat:@"%@%@", @(@encode(id)), NSStringFromSelector(@encode(SEL))];
        methodSignature = [NSMethodSignature signatureWithObjCTypes:types.UTF8String];
    }
    return methodSignature;
}

- (void)forwardInvocation:(NSInvocation *)anInvocation
{
    id target = self.target;
    if (target) {
        [anInvocation invokeWithTarget:target];
    }
}

@end
```

8W9aG replied:
Hmmm can we not invalidate the garbage collection scheduler in the cache on dealloc and call it a day?

rastersize replied:
@8W9aG That’s what I proposed in “solution one”. Added with removing this `-dealloc` implementation and documenting the requirement.

rastersize replied:
@8W9aG However that is slightly brittle as it depends on humans reading documentation to avoid bugs. While the second solution is a technical solution that doesn’t require human effort nor relies on said humans to read and remember the documentation.

ChocoChipset replied:
Solution 1, is already in place: `SPTPersistentCache` unschedules the garbage collector on `-dealloc`.

However, I should add a 'don't forget to unschedule the garbage collector' to avoid retain cycles'
 note in the class documentation, and remove the custom logic in `dealloc`, indeed there's nothing useful on it given this approach.

ChocoChipset replied:
Added warnings.

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,141 @@
+    }
+}
+
+#pragma mark -
```
Missing text for this pragma mark :)

ChocoChipset replied:
"intentionally left blank", just to give some space while browsing the selectors in the navigator.

rastersize commented on Sources/SPTPersistentCacheTypeUtilities.h:
```diff
@@ -29,3 +30,13 @@
+ *  @param debugMessage The debug message.
+ *  @param debugCallback The callback block to execute safely.
+ */
+void SPTPersistentCacheSafeDebugCallback(NSString *debugMessage,
```
This is a bit of a nitpick but `SPTPersistentCacheSafeDebugCallback()` is not a _type utility_. Which this header is for according to its name. How about moving this function to its own file? E.g. `SPTPersistentCacheDebugging.h`, `SPTPersistentCacheDebugUtilities.h` or something like that?

rastersize commented on Tests/SPTPersistentCacheGarbageCollectorSchedulerTests.m:
```diff
@@ -48,13 +55,14 @@ - (void)pruneBySize
-@property (nonatomic, strong) SPTPersistentCacheTimerProxy *timerProxy;
+@interface SPTPersistentCacheGarbageCollectorSchedulerTests : XCTestCase
+@property (nonatomic, strong) SPTPersistentCacheOptions *options;
+@property (nonatomic, strong) SPTPersistentCacheGarbageCollectorScheduler *timerProxy;
```
`*timerProxy` :point_right: `*garbageCollectorScheduler` (or if class is renamed, just `*garbageCollector`).

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.h:
```diff
@@ -21,8 +21,9 @@
+@class SPTPersistentCacheOptions;
 
-@interface SPTPersistentCacheTimerProxy : NSObject
+@interface SPTPersistentCacheGarbageCollectorScheduler : NSObject
```
Been thinking, this name implies that the class has two jobs:
- Garbage collection
- Scheduling

How about we rename it to just `SPTPersistentCacheGarbageCollector`? As running the garbage collection is its primary responsibility.

Then the `-scheduleGarbageCollection` could be named just `-schedule` and still be clear while also being more succinct. The same for `-unscheduleGarbageCollection` which could be just `-unschedule`.

What do you think @iOSCowboy?

ChocoChipset replied:
Great suggestion!

ChocoChipset replied:
Renamed.

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.h:
```diff
@@ -35,14 +36,28 @@
  *  @param cache Persistent Cache that will be used for garbage collection operations.
  *  @param queue Dispatch queue where the operations will take place.
  */
 - (instancetype)initWithDataCache:(SPTPersistentCache *)cache
```
How about we remove the “Data” part from the method name? As well as from the property.

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.h:
```diff
@@ -35,14 +36,28 @@
+/**
  *  Initializes the timer proxy on a specific queue using a specific data cache.
  *  
  *  @param cache Persistent Cache that will be used for garbage collection operations.
```
Add parameter documentation for `options`

ChocoChipset replied:
😎

rastersize commented on Sources/SPTPersistentCacheGarbageCollectorScheduler.m:
```diff
@@ -0,0 +1,141 @@
+    _options = options;
+    _dataCache = dataCache;
+    _queue = queue;
+    _debugOutput = options.debugOutput;
```
My comment previously wasn’t that the block was being copied (which was correct from a memory management perspective). Rather, why do we need an extra reference to it at all? We could already do `self.options.debugOutput`? How about we just remove this duplicate the property?

rastersize commented:
You can find lines that aren’t tested (which is why Codecov failed the PR, it decreased test coverage) at
- https://codecov.io/github/spotify/SPTPersistentCache?ref=e2f0f4806dc48022d3812124567cc9f80f5ea4c6 and
- https://codecov.io/github/spotify/SPTPersistentCache/compare/5f68750...e2f0f48

Commit:
Rename timerProxy to garbageCollectorScheduler in SPTPersistentCacheGarbageCollectorSchedulerTests

Commit:
Document 'options' argument in initializer of SPTPersistentCacheGarbageCollectorScheduler

Commit:
Remove unecesary reference to debugCallback block in SPTPersistentCacheGarbageCollectorScheduler

Commit:
Document requirement of unscheduling the garbage collector to dealloc it

Commit:
Move debug helper function to its own DebugUtilities class

Commit:
Remove 'scheduler' prefix from garbage collector

Commit:
Drop legacy 'data' word from SPTPersistentCacheGarbageCollector initializer

Commit:
Drop 'garbageCollection' suffix from methods

Commit:
Merge branch 'master' of github.com:spotify/SPTPersistentCache into timer-refactor

Commit:
Avoid executing the garbage collector operations outside the main thread

Commit:
Add Debug Utilities class to framework project

ChocoChipset committed:
Merge branch 'master' into timer-refactor

Commit:
Merge branch 'master' of github.com:spotify/SPTPersistentCache into timer-refactor

Commit:
Fix compiling error in SPTPersistentCacheGarbageCollectorTests

Commit:
Rename 'dataCache' property to 'cache' in Garbage Collector

rastersize commented:
:+1: :deciduous_tree: 

Thank you :bow:

Commit:
Compare queue labels as strings and not by comparing their pointers
